src/highlight/highlightUtil.ts

```diff
@@ -58,3 +58,65 @@ export const centerHighlight = (shapes: Shape[]): Position => {
 export function isHighlight(annotation: Annotation): annotation is AnnotationHighlight {
     return annotation?.target?.type === Type.highlight;
 }
+
+export const dedupeRects = (rects: Shape[], threshold = 0.1): Shape[] => {
+    const dedupedRects: Shape[] = [];
+
+    rects.forEach(curr => {
+        const prev = dedupedRects.pop();
+        // empty list, push current
+        if (!prev) {
+            dedupedRects.push(curr);
+            return;
+        }
+
+        const heightThreshold = Math.max(
+            Math.floor(prev.height * threshold + 0.5),
+            Math.floor(curr.height * threshold + 0.5),
+            1,
+        );
+
+        // the same rect, push the larger one
+        if (prev.x === curr.x && prev.width === curr.width && Math.abs(prev.y - curr.y) <= heightThreshold) {
+            dedupedRects.push(prev.height > curr.height ? prev : curr);
+            return;
+        }
+
+        // different rects, push both
+        dedupedRects.push(prev);
+        dedupedRects.push(curr);
+    });
+
+    return dedupedRects;
+};
+
+export const groupByRow = (shapes: Shape[]): Record<number, Shape[]> => {
+    const rows: Record<number, Shape[]> = {};
+    shapes.forEach(shape => {
+        const { y } = shape;
+        const roundedY = Math.floor(y + 0.5);
```

Comment:
This seems to be doing more than rounding? How/why did we arrive at `0.5` for this value? Does it scale across different font sizes and line heights?

Reply:
I actually want to use `Math.round`... (`Math.floor(y + 0.5) === Math.round(y)`)